### PR TITLE
Web SDK - Developer Identity not persisting across pageviews & sessions

### PR DESCRIPTION
**Pull request by jwangler-branch**
Before I create a PR to our public master I want to finish an internal Pull Request. 

PR: https://branch.atlassian.net/secure/RapidBoard.jspa?rapidView=336&projectKey=JOUR&modal=detail&selectedIssue=INTENG-7557

On creating a new tab or refreshing the current page in the browser, the developer identity doesn't survive. Therefore the setIdentity() method has to be called to set a new one. The goal was to keep the id alive and reuse it on refresh / new tap. 

Change Log

- Add field identity to local and session storage 
- The identity will be reused from the storage on browser refresh and create a new tab.

### CODE REVIEW

File: src/2_session.js [hunk 1]
```diff
@@ -50,6 +50,31 @@ session.update = function(storage, newData) {
 	storage.set('branch_session', data);
 };
 
+/**
+ * Patch a field in localStorage or sesstionStorage or both.
```
Comment (jdee) [inline src/2_session.js]: Spelling in comment: sessionStorage.

Comment (jwangler-branch) [reply src/2_session.js]: Thanks 👍

**Comment by rubinsingh** [comment]
@jwangler-branch could you please remove build.js and build.min.js from the PR? Only source code changes are required in websdk prs.

File: src/2_session.js [hunk 2]
```diff
@@ -50,6 +50,31 @@ session.update = function(storage, newData) {
...
+	const merge = (source, patch) =>{
+		return utils.merge(source, patch);
+	}
+
+	const session = storage.get('branch_session', false) || {}
+	if(session){
+		storage.set('branch_session', goog.json.serialize(merge(session, data)));		
+	}
+
+	if(druable){
```
Comment (rubinsingh) [inline src/2_session.js]: Can we change this variable name to updateLocalStorage?

**Comment by rubinsingh** [comment]
@jwangler-branch could you please manually add either identity (v1 endpoints) or developer_identity (v2 endpoints) to those requests that are not wrapped by the defaults object?

File: src/2_session.js (hunk 2 above)
Comment (jwangler-branch) [reply src/2_session.js]: 👍

**Comment by rubinsingh** [comment]
@jwangler-branch could you remove build.min.js and build.js from your PR those are automatically generated

File: src/2_session.js [hunk 3]
```diff
@@ -46,16 +46,41 @@ session.update = function(storage, newData) {
...
+ * @param {boolean=} druable
+ */
+session.patch = function(storage, data, updateLocalStorage){
+
+	const merge = (source, patch) =>{
+		return utils.merge(source, patch);
+	}
+
+	const session = storage.get('branch_session', false) || {}
+	if(session === true){
```
Comment (rubinsingh) [inline src/2_session.js]: Do we need this check if session is always going to be defined?

File: src/2_resources.js [hunk 4]
```diff
@@ -129,7 +129,8 @@ resources.linkClick = {
 		"link_url": validator(true, validationTypes.STRING)
 	},
 	params: {
-		"click": validator(true, validationTypes.STRING)
+		"click": validator(true, validationTypes.STRING),
+		"identity": validator(true, validationTypes.STRING)
```
Comment (ahmednawar) [inline src/2_resources.js]: don't think we need to send identity for link clicks

**Comment by ahmednawar** [comment]
overall looks good, we just need to ensure it passes the regression testing and fixes the original bug reported.

File: src/6_branch.js [hunk 5]
```diff
@@ -553,13 +590,14 @@ Branch.prototype['init'] = wrap(
 							"current_url": utils.getCurrentUrl(),
 							"screen_height": utils.getScreenHeight(),
 							"screen_width": utils.getScreenWidth()
+							//TOO
```
Comment (rubinsingh) [inline src/6_branch.js]: Is there a reason for this comment?

File: src/2_session.js (hunk 3 above)
Comment (rubinsingh) [reply src/2_session.js]: Or did you mean to do if (Object.keys(session).length > 0) ?

**Comment by rubinsingh** [comment]
👍

**Comment by jwangler-branch** [comment]
Web SDK - Developer Identity not persisting across pageviews & sessions

**Pull request merged**